cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,492 @@
+            self.solver_is_initialized=True
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            self.xct_solver.setOption(kw,str(val))
+
+        # set assumptions
+        if assumptions is not None:
+            assert(all(is_bool(v) for v in assumptions))
+            assump_vals = [int(not isinstance(v, NegBoolView)) for v in assumptions]
+            assump_vars = [self.solver_var_no_num(v._bv if isinstance(v, NegBoolView) else v) for v in assumptions]
+            self.assumption_dict = {xct_var: (xct_val,cpm_assump) for (xct_var, xct_val, cpm_assump) in zip(assump_vars,assump_vals,assumptions)}
+            self.xct_solver.setAssumptions(assump_vars, assump_vals);
+            # NOTE: setAssumptions does not clear previous assumptions, is this ok?
+
+
+        # call the solver, with parameters
+        start = time.time()
+        my_status = self.xct_solver.runFull(self.has_objective, time_limit if time_limit is not None else 0)
+        end = time.time()
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = end - start
+
+        # translate exit status
+        if my_status == 0: # found unsatisfiability
+            if self.xct_solver.hasSolution():
+                self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+            else:
+                self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif my_status == 1: # found solution, but not optimality proven
+            assert self.xct_solver.hasSolution()
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif my_status == 2: # found inconsistency
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN # TODO
+        elif my_status == 3: # found timeout
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN # TODO
+        else:
+            raise NotImplementedError(my_status)  # a new status type was introduced, please report on github
+
+        return self.getSolAndObj()
+
+    def solveAll(self, display=None, time_limit=None, solution_limit=None, **kwargs):
+        # TODO: test this function
+        """
+            Compute all solutions and optionally display the solutions.
+
+            This is the generic implementation, solvers can overwrite this with
+            a more efficient native implementation
+
+            Arguments:
+                - display: either a list of CPMpy expressions, OR a callback function, called with the variables after value-mapping
+                        default/None: nothing displayed
+                - time_limit: stop after this many seconds (default: None)
+                - solution_limit: stop after this many solutions (default: None)
+                - any other keyword argument
+
+            Returns: number of solutions found
+        """
+
+        assert not self.has_objective # TODO: is this true?
+
+        if not self.solver_is_initialized:
```

Comment:
if this is in the constructor, you avoid having two places where you do this

Reply:
It's needed to initialize the objective, which may not be ideal, but I'm not going to refactor that part of Exact soon.